third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -1408,11 +1426,35 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
                 // The type of the assign is Unit
                 let result_ty = self.check_type(&loc, &Type::unit(), expected_type, context);
                 let id = self.new_node_id_with_type_loc(&result_ty, &loc);
-                ExpData::Assign(id, lhs, rhs.into_exp())
+                let lhs_ty = self.env().get_node_type(lhs.node_id());
+                let rhs_ty = self
+                    .subs
+                    .specialize(&self.env().get_node_type(rhs.node_id()));
+                let mut rhs = rhs.into_exp();
+                // Insert freeze for rhs of the assignment
+                if lhs_ty.is_tuple() && rhs_ty.is_tuple() {
+                    if let Pattern::Tuple(_, lhs_tys) = &lhs {
+                        if let Type::Tuple(rhs_tys) = &rhs_ty {
+                            let lhs_tys = lhs_tys
+                                .iter()
+                                .map(|pat| self.get_node_type(pat.node_id()))
+                                .collect_vec();
+                            if let ExpData::Call(_, Operation::Tuple, _) = rhs.as_ref() {
```

Comment:
Do this test for `ExpData::Call` first if there's any chance of it not being true.  Rust compiler might be able to reorder the tests, but I wouldn't take that chance.

Reply:
done